### PR TITLE
partial #15842, merge #15849, #17342, #18710: Add local thread pool to CCheckQueue

### PR DESCRIPTION
**Pull request by kittywhiskers**
# Overview

~~**This pull request presupposes the merger of https://github.com/dashpay/dash/pull/4190 into `develop`.** Only the last four commits are meant to be reviewed.~~ This pull request adds a local thread pool to CCheckQueue in preparation for more threading refactors. ~~To actually compare `checkqueue` with `tlocks`, go [here](https://github.com/kittywhiskers/dash/compare/tlocks...kittywhiskers:checkqueue)~~

## Motivation

During initial preparation for https://github.com/bitcoin/bitcoin/pull/18242 ("Add BIP324 encrypted p2p transport de-/serializer" by jonasschnelli) and the backporting of https://github.com/bitcoin/bitcoin/commit/78c312c983255e15fc274de2368a2ec13ce81cbf, a difference in class and function names were noticed and that in turn lead to a series of pull requests the lead us to this one.

The motivation is to reduce future merge conflicts when backporting pull requests from Bitcoin Core.

## Contents

 #15849, #17342, #18710

### Partials

 #15842

# Disclosures

* **This is a work in progress and needs external review**. Dash-specific changes have not been tested, only compilation and unit test success have been ensured. 
* Contains squashed patches b87b772a57c30e2393284414291f99f2d2bf81f9, c6ced95b39027ff95308e2b563606e425639ae9c, 589ffcfee4db1241938510487f022a528c76e68b and b071fe038312ec915f0e6bb6e42e7c27267d3b01 by UdjinM6

### CODE REVIEW

**Comment by UdjinM6** [comment]
Pls see https://github.com/UdjinM6/dash/commits/pr4191

**Comment by UdjinM6** [comment]
Looking at the diff now I think I messed up in my fixes a bit 🙈 , pls see https://github.com/UdjinM6/dash/commit/b071fe038312ec915f0e6bb6e42e7c27267d3b01

File: src/logging.cpp [hunk 1]
```diff
@@ -244,43 +247,29 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
...
+    str_prefixed = LogTimestampStr(str_prefixed);
 
-    if (!str.empty() && str[str.size()-1] == '\n')
-        m_started_new_line = true;
-    else
-        m_started_new_line = false;
+    m_started_new_line = !str.empty() && str[str.size()-1] == '\n';
 
-    if (m_buffering) {
+    if (m_fileout) {
```
Comment (UdjinM6) [inline src/logging.cpp:264]:
```suggestion
    if (m_buffering) {
```

File: doc/release-notes-15849.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Thread names in logs
+--------------------
+
+On platforms supporting `thread_local`, log lines can be prefixed with the name
+of the thread that caused the log. To enable this behavior, use
+`-logthreadnames=1`.
```
Comment (PastaPastaPasta) [inline doc/release-notes-15849.md:6]: should drop this

File: src/logging.cpp [hunk 3]
```diff
@@ -6,6 +6,9 @@
 #include <logging.h>
 #include <util.h>
 #include <utiltime.h>
+#include <utilthreadnames.h>
```
Comment (PastaPastaPasta) [inline src/logging.cpp:9]: this should be above util time

Comment (kittywhiskers) [reply src/logging.cpp:8]: Resolved

File: doc/release-notes-15849.md (hunk 2 above)
Comment (kittywhiskers) [reply doc/release-notes-15849.md]: Resolved

**Comment by UdjinM6** [comment]
LGTM, pls squash c773696 into 187f473 (pr 15849)

File: src/httpserver.cpp [hunk 4]
```diff
@@ -9,6 +9,7 @@
 #include <compat.h>
 #include <util.h>
 #include <utilstrencodings.h>
+#include <utilthreadnames.h>
```
Comment (PastaPastaPasta) [inline src/httpserver.cpp:12]: should be moved into util folder depending on order of merges

**Review by PastaPastaPasta — COMMENTED** [review]
Looks good, besides udjin squash and util stuff will require #4197 to be updated

File: src/httpserver.cpp (hunk 4 above)
Comment (kittywhiskers) [reply src/httpserver.cpp:12]: Would suggest merging the refactor first, then I'll replay this PR on top of it and put it in the util directory

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Review by UdjinM6 — APPROVED** [review]
slightly tested ACK

**Pull request merged**